### DISCUSSION TITLE
(Auto)-sync content directly from Roam

### DISCUSSION THREAD

**Discussion by Stvad** [roam-garden/i#28] [Feature requests & Ideas] [answered]
I think it can be a great experience to just have a `Publish` button inside Roam (based on [[roam/js]]) as an entry point. 
When user presses this button - the roam/js client would filter things according to your configuration and send them to the [[API]] that would do the publishing.

It can also run on a regular basis for automatic publishing (as long as you have Roam tab open)

Roadmap: https://github.com/roam-garden/i/projects/1#card-56652560

### Configuration
**Option 1 (preferred)**: Store configuration in a special page inside Roam. I know of several 

**Option 2**: Have configuration stored inside the roam/js snippet people would copy 

### Filtering
I'd like to re-use the logic from https://github.com/roam-garden/roam-export to provide unified experience for "upload json" and "publish from Roam" use-cases.  That package operates on Roam JSON export though. So a first step before I can reuse it is to generate a JSON-export like data-structure myself. 

[[David Vargas]] is doing something similar in https://github.com/dvargas92495/generate-roam-site/commit/df634c8350d45fd45553da1f87183bf465364869

Other interesting resource on that https://davidbieber.com/snippets/2020-12-22-datalog-queries-for-roam-research/

### Other possible features
- This may also enable extracting information that is not currently encoded in JSON export (#19)
- Make the "make-public" marker subtly change the background of the page inside Roam (or add some other indicator). To make it obvious that the page you're editing is shared. 

### Roam Extension name options
- Arborist
- Plantsman 
- Gardner (conflicts with internal things)

---
![vo6ohmcfSc](https://user-images.githubusercontent.com/81347/118224015-2ad90a00-b437-11eb-94ed-481c9cd2b546.png)

**Comment by kalmir** [comment]
That would indeed be great.

**Reply by Stvad** [reply]
you should upvote it then ;)

**Comment by Stvad** [comment, marked as answer]
Upon initial research I figured out a way to do this in an easier then expected way! I hope to have a first test version by the end of the week! LMK if you want to be an early tester!

**Reply by Stvad** [reply]
Ok. It's ready for testing!

You'd need to install the following roam/js block to your db (& make sure to agree for it to run. see https://roamstack.com/how-install-roam-plugin/ for general roam/js plugin installation instructions):
```
- {{[[roam/js]]}}
    - ```javascript
let id = "roam-garden-plantsman"
let existingRg = document.getElementById(id)
if (!existingRg) {
    const extension = document.createElement("script")
    extension.src = "https://resources.roam.garden/js/plantsman.min.js"

    extension.id = id
    extension.async = true
    extension.type = "text/javascript"
    document.getElementsByTagName("head")[0].appendChild(extension)
}```

and then navigate to `roam/garden` page to see the UI (create it if it does not exist)

**Reply by Stvad** [reply]
A few things to do before this is fully ready:
- [x] styling
- [x] better success indication (checkmark/redirect to special page?)
- [x] display username (maybe a header with it, logout button, and link to Roam Garden)
- [x] handle lack of subscription in a better way then just crashing 😛 
- [x] a way to make the publishing page more discoverable (add an icon in the toolbar?)
  - Did a command pallet for now
- [x] create `roam/garden` if it does not exist?
- [ ] css entry validation. it seems it's a lot easier to make an error in live editor 😅 
  - Did an iteration on this with https://github.com/csstree/validator it misses some issues though :( (https://github.com/csstree/validator/issues/21 https://github.com/csstree/validator/issues/20)
  - PostCSS can give me a yes/no result on validity, but not sure if I can get a list of all errors https://github.com/postcss/postcss/discussions/1555

Separately things to do on the main website to facilitate setup:
- [ ] Update the upload form to use new/improved components
- [ ] Initial page should have a link to setup instructions at help.roam.garden & place to upload JSON to initiate upload otherwise
- [ ] need to have those instructions at help.roam.garden
  - [ ] they should have a video/gif on installing the roam/js and seeing the ui
- [ ] Update FAQ to indicate improved data handling

**Reply by Stvad** [reply]
I think this generally ready to use. Find more detailed setup instructions at https://help.roam.garden/?stackedPages=%2FRoam-Garden-plugin-setup

**Reply by JayFarei** [reply]
hi @Stvad  - can I confirm this approach creates a filtered export locally before sending it to roam.garden's APIs? 

Is there a way to add a verification of the export in the middle to reassure the user that no information is leaked? 

I am experimenting with it right now and I much prefer this method than the full export & server side filtering, however I am a still a bit anxious of potential data leakage as the public content lives alongside private content in my graph.

**Reply by Stvad** [reply]
Hey @JayFarei! Yes that is correct all the filtering happens locally 🙂  I note on that in the FAQ section ;) 
And the plugin code is open-source at https://github.com/roam-garden/plantsman

Also, the upload process on https://roam.garden/ was updated to also do filtering locally, so now only the data explicitly marked public makes it to the API regardless of the upload method

>Is there a way to add a verification of the export in the middle to reassure the user that no information is leaked?

There is nothing like that atm. Can you elaborate on how you see this interaction working? 

>I am experimenting with it right now and I much prefer this method than the full export & server side filtering, however I am a still a bit anxious of potential data leakage as the public content lives alongside private content in my graph.

Makes sense, can you elaborate on what kind of things would you find reassuring for this? :)

**Reply by JayFarei** [reply]
Not sure exactly how to do that - Obsidian publish I think offers a UI that shows which pages will be published I believe.
You could implement a similar step in your Roam/js publication workflow so that a user can spot mistakes prior publication.

Alternatively you could also offer the option to save on the local machine the exported markdowns (I imagine you do that server side atm for the gatsby template?) & a way to keep those in sync with your gatsby site agent? (that makes your service  also somewhat Roam agnostic => e.g. logseq)

**Reply by Stvad** [reply]
>Not sure exactly how to do that - Obsidian publish I think offers a UI that shows which pages will be published I believe.

Hmm, it can be a lot of pages though 🤔. Would be interested to see how it looks in Obsidian publish. 
Will experiment with doing something like that, should probably create a separate request for it though 

>Alternatively you could also offer the option to save on the local machine the exported markdowns (I imagine you do that server side atm for the gatsby template?) & a way to keep those in sync with your gatsby site agent? (that makes your service also somewhat Roam agnostic => e.g. logseq)

Not quite sure this addresses the issue in question 🙂. And yes this happens on the backend, though what's generated is not markdown, but MDX using & custom components. + There is a lot of post-processing done on generated code, so it by itself is not very useful =\

**Reply by Stvad** [reply]
started https://github.com/roam-garden/i/discussions/63 for confirmation dialog discussion
